### ISSUE TITLE
Sort order with multiple keys

### ISSUE THREAD

**Issue by luksch** [lodash/lodash#414]
Hi, I am a in awe of the power of lodash. For my project it has been of invaluable help so far. However, now I am stuck with a problem with which lodash seems to fall a bit short.

Here is the problem:

I have an array of objects with the same properties, say 

```
[{name:"barney", age: 38},
 {name:"fred", age: 40},
 {name:"fred", age: 12}]
```

I want to sort the array by name in descending order, but age should be in ascending order. 

lodash supports multiple sorting keys by the "pluck" syntax, but only can deal with ascending sorting this way. Is there a way that the sortBy function is altered that it also accepts an array of objects that define the property name (key) and the order direction? I would love to do:

```
mySortedArr = _.sortby(
    unsortedArr, [
        {propertyName:"name", order:"desc"}, 
        {propertyName:"age", order:"asc"}
    ]);
```

Any chance something like that enters the library?

**Comment by darthsteedious** [comment]
+1 - would love a _.sortBy(myArray, ['foo', '-bar']) syntax.

**Comment by ghost1face** [comment]
+1

**Comment by jdalton** [comment]
I'd dig a proposed implementation to go along with the +1's.
In 3.0 we're separating `_.sortBy` out into `_.sortByAll` to handle multiple properties. This lets use keep `_.sortBy` simple.

**Comment by ItsLeeOwen** [comment]
+1

**Comment by CTAPbIuMABP** [comment]
You always can add _.sortBySome

**Comment by wrummler** [comment]
I'd love to see `sortByAll` handle optionally-specified per-criterion descending sort order. (Btw, gotta mention it... Like many others, I am in love with the performance and expressiveness of lodash. Thanks for your awesome work.)

Perhaps a signature something like `sortByAll(collection, props, order)` for usage like `_.sortByAll(myArray, ['foo', 'bar'], [false, true])`, to borrow from @octref 's example?

Copying & adapting `compareMultipleAscending` to `compareMultiple` seems straightforward (to me at least, relatively new to JS and possibly unaware of some pitfalls):

    /**
     * @param {Array} order Each truthy element indicates descending sort order for the corresponding criterion.
     */
    function compareMultiple(object, other, order) { // diff
        var index = -1,
            objCriteria = object.criteria,
            othCriteria = other.criteria,
            length = objCriteria.length;
    
        while (++index < length) {
            var result = baseCompareAscending(objCriteria[index], othCriteria[index]);
            if (result) {
                return order[index] ? -result : result; // diff
            }
        }
    
        return object.index - other.index;
    }

What I like about this approach, using an order-indicating array side-by-side with the criteria array, is that it robustly (I think) defaults to all-ascending when left unspecified. It also doesn't restrict the names of properties like the `+`/`-` prefix suggestion does.

The thing I couldn't figure out is *correctly* modifying `sortByAll` to "carry" `order` from the argument list through to e.g. `return baseSortBy(result, curryRight(compareMultiple)(order))`.

I get tripped up determining whether the iteratee-check area needs any changes, and then also around the call to `baseFlatten` which might, to continue from the example, end up assigning `['foo', 'bar', false, true]` to `props` if I do something dumb. I don't yet grok things well enough to confidently resolve this.

Anyway, it'd be great to have something like this working for `sortByAll`. Thanks again.

**Comment by jdalton** [comment]
I'd prefer an options object, maybe `dir` with a `1` or `-1`.

**Comment by ItsLeeOwen** [comment]
Allow props array to optionally contain objects with dir as a value?

```javascript
var users = [
  { 'user': 'barney', 'age': 36 },
  { 'user': 'fred',   'age': 40 },
  { 'user': 'barney', 'age': 26 },
  { 'user': 'fred',   'age': 30 }
];

// default to standard direction
_.sortByAll(myArray, ['user', 'age'])

or

// define direction
_.sortByAll(myArray, [{user: 1}, {age: -1}])
```

**Comment by jdalton** [comment]
Crawling keys is not ideal. It'd be better to have `[{ 'key': 'user', 'dir': 1 },{ 'key': 'age', 'dir': -1 }]`

**Comment by octref** [comment]
I think for the sake of consistency, this might be better:
```
_.sortByAll(myArray, ['user', 'age'])
_.sortByAll(myArray, {user: 1, age: -1})
_.sortByAll(myArray, comp)
```
iteratee / predicate style, what do you think @jdalton ?

**Comment by jdalton** [comment]
What's `comp` (a function?)

**Comment by jdalton** [comment]
I'll accept a PR for `_.sortByAll(myArray, {user: 1, age: -1})`.

**Comment by octref** [comment]
Sounds good. I guess I'll take a stab at that tonight. 
Forget comp it should belong to sortBy.

**Comment by ItsLeeOwen** [comment]
Isn't the problem with using a sort object javascript' _.sortByAll(myArray, {user: 1, age: -1})

4.3.3 Object
An object is a member of the type Object. It is an unordered collection of properties each of which contains a primitive value, object, or function. A function stored in a property of an object is called a method.

**Comment by octref** [comment]
Right. I guess we'd have to type the long 
`[{ 'key': 'user', 'dir': 1 }, { 'key': 'age', 'dir': -1 }]` then :-(.
And I feel `'dir': 'asc'` and `'dir': 'desc'` looks better, but that's just my opinion.

**Comment by ItsLeeOwen** [comment]
Maybe to streamline ui implementation consider "reverse: true/false" instead of dir.
```javascript
_.sortByAll(myArray, [{key: "user"}, {key: "age", reverse: true}])

$scope.sort = [{key: 'user'}, {key: 'age', reverse: true}]
```
```html
<ng-repeat="option in sort">
  <caret ng-click="option.reverse = !option.reverse"></caret>
  <label>{{::option.key}}</label>
</ng-repeat>
```

**Comment by jdalton** [comment]
since sorting is ascending by default how about a `desc` that's either truthy or not.

**Comment by ItsLeeOwen** [comment]
^^  +1 for desc: true/false

**Comment by octref** [comment]
I'm good with `desc` true/false.
So it'll look like this:
```
_.sortByAll(users, [
  { 'key': 'user', 'desc': true },
  { 'key': 'age'}, // Default to 'desc': false
  { 'key': 'gender', 'desc': false } // You can also say that explicitly
]);


```

**Comment by jdalton** [comment]
keep in mind it would only be needed when different than the default. So regular keys could be used for the common case.

**Comment by NeevShah** [comment]
Hello @jdalton Is there any way to pass values with keys in _.sortByAll() ? Can I pass boolean as a value?

**Comment by jdalton** [comment]
>  Is there any way to pass values with keys in `_.sortByAll()`? Can I pass boolean as a value?

I'm not sure what you're trying to accomplish. There's [_.sortByOrder](https://lodash.com/docs#sortByOrder) which allows specifying booleans for sort order direction.

**Comment by NeevShah** [comment]
I'm responsible for upgrading lodash to 3.8 version from 2.4.1. Where I came to this code where I thought I can improve code. 
So my use case is I want to sort array of object by two properties with values. And Following is current code and code I did try to overcome.

Example: var arr = [{"name" :  "test' , type : "1",isActive :true},{"name" :  "tea', type : "2" , isActive :false},{"name" :  "tea' , type : "1", isActive :true}]; 

Now I want to sort first with {"type" : "1" } and then {isActive : false}.. In 2.4.1 I was doing  like 
_.sortBy(_.sortBy(arr, {"type" : 1}), {"isActive" : false}));

Can I use _.sortByAll to make it readable? I was trying 
_.sortByAll(arr, {"type" : 1, "isActive" : false}); but this didn't work for me..
